webgpu.h

```diff
@@ -0,0 +1,1025 @@
+    WGPU_TEXTURE_FORMAT_RGBA32_SINT = 0x00000028,
+    WGPU_TEXTURE_FORMAT_DEPTH32_FLOAT = 0x00000029,
+    WGPU_TEXTURE_FORMAT_DEPTH24_PLUS = 0x0000002A,
+    WGPU_TEXTURE_FORMAT_DEPTH24_PLUS_STENCIL8 = 0x0000002B,
+    WGPU_TEXTURE_FORMAT_BC1_RGBA_UNORM = 0x0000002C,
+    WGPU_TEXTURE_FORMAT_BC1_RGBA_UNORM_SRGB = 0x0000002D,
+    WGPU_TEXTURE_FORMAT_BC2_RGBA_UNORM = 0x0000002E,
+    WGPU_TEXTURE_FORMAT_BC2_RGBA_UNORM_SRGB = 0x0000002F,
+    WGPU_TEXTURE_FORMAT_BC3_RGBA_UNORM = 0x00000030,
+    WGPU_TEXTURE_FORMAT_BC3_RGBA_UNORM_SRGB = 0x00000031,
+    WGPU_TEXTURE_FORMAT_BC4_R_UNORM = 0x00000032,
+    WGPU_TEXTURE_FORMAT_BC4_R_SNORM = 0x00000033,
+    WGPU_TEXTURE_FORMAT_BC5_RG_UNORM = 0x00000034,
+    WGPU_TEXTURE_FORMAT_BC5_RG_SNORM = 0x00000035,
+    WGPU_TEXTURE_FORMAT_BC6H_RGB_UFLOAT = 0x00000036,
+    WGPU_TEXTURE_FORMAT_BC6H_RGB_SFLOAT = 0x00000037,
+    WGPU_TEXTURE_FORMAT_BC7_RGBA_UNORM = 0x00000038,
+    WGPU_TEXTURE_FORMAT_BC7_RGBA_UNORM_SRGB = 0x00000039,
+    WGPU_TEXTURE_FORMAT_FORCE32 = 0x7FFFFFFF
+} wgpuTextureFormat;
+
+typedef enum {
+    WGPU_TEXTURE_VIEW_DIMENSION_1D = 0x00000000,
+    WGPU_TEXTURE_VIEW_DIMENSION_2D = 0x00000001,
+    WGPU_TEXTURE_VIEW_DIMENSION_2D_ARRAY = 0x00000002,
+    WGPU_TEXTURE_VIEW_DIMENSION_CUBE = 0x00000003,
+    WGPU_TEXTURE_VIEW_DIMENSION_CUBE_ARRAY = 0x00000004,
+    WGPU_TEXTURE_VIEW_DIMENSION_3D = 0x00000005,
+    WGPU_TEXTURE_VIEW_DIMENSION_FORCE32 = 0x7FFFFFFF
+} wgpuTextureViewDimension;
+
+typedef enum {
+    WGPU_VERTEX_FORMAT_UCHAR2 = 0x00000000,
+    WGPU_VERTEX_FORMAT_UCHAR4 = 0x00000001,
+    WGPU_VERTEX_FORMAT_CHAR2 = 0x00000002,
+    WGPU_VERTEX_FORMAT_CHAR4 = 0x00000003,
+    WGPU_VERTEX_FORMAT_UCHAR2_NORM = 0x00000004,
+    WGPU_VERTEX_FORMAT_UCHAR4_NORM = 0x00000005,
+    WGPU_VERTEX_FORMAT_CHAR2_NORM = 0x00000006,
+    WGPU_VERTEX_FORMAT_CHAR4_NORM = 0x00000007,
+    WGPU_VERTEX_FORMAT_USHORT2 = 0x00000008,
+    WGPU_VERTEX_FORMAT_USHORT4 = 0x00000009,
+    WGPU_VERTEX_FORMAT_SHORT2 = 0x0000000A,
+    WGPU_VERTEX_FORMAT_SHORT4 = 0x0000000B,
+    WGPU_VERTEX_FORMAT_USHORT2_NORM = 0x0000000C,
+    WGPU_VERTEX_FORMAT_USHORT4_NORM = 0x0000000D,
+    WGPU_VERTEX_FORMAT_SHORT2_NORM = 0x0000000E,
+    WGPU_VERTEX_FORMAT_SHORT4_NORM = 0x0000000F,
+    WGPU_VERTEX_FORMAT_HALF2 = 0x00000010,
+    WGPU_VERTEX_FORMAT_HALF4 = 0x00000011,
+    WGPU_VERTEX_FORMAT_FLOAT = 0x00000012,
+    WGPU_VERTEX_FORMAT_FLOAT2 = 0x00000013,
+    WGPU_VERTEX_FORMAT_FLOAT3 = 0x00000014,
+    WGPU_VERTEX_FORMAT_FLOAT4 = 0x00000015,
+    WGPU_VERTEX_FORMAT_UINT = 0x00000016,
+    WGPU_VERTEX_FORMAT_UINT2 = 0x00000017,
+    WGPU_VERTEX_FORMAT_UINT3 = 0x00000018,
+    WGPU_VERTEX_FORMAT_UINT4 = 0x00000019,
+    WGPU_VERTEX_FORMAT_INT = 0x0000001A,
+    WGPU_VERTEX_FORMAT_INT2 = 0x0000001B,
+    WGPU_VERTEX_FORMAT_INT3 = 0x0000001C,
+    WGPU_VERTEX_FORMAT_INT4 = 0x0000001D,
+    WGPU_VERTEX_FORMAT_FORCE32 = 0x7FFFFFFF
+} wgpuVertexFormat;
+
+typedef enum {
+    WGPU_BUFFER_USAGE_BIT_NONE = 0x00000000,
```

Comment:
Because AFAIK that's the best practice, for example Vulkan has this (although now I realize that they use the VkFlags type):

```c
typedef enum VkImageUsageFlagBits {
    VK_IMAGE_USAGE_TRANSFER_SRC_BIT = 0x00000001,
    VK_IMAGE_USAGE_TRANSFER_DST_BIT = 0x00000002,
    VK_IMAGE_USAGE_SAMPLED_BIT = 0x00000004,
    VK_IMAGE_USAGE_STORAGE_BIT = 0x00000008,
    VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT = 0x00000010,
    VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT = 0x00000020,
    VK_IMAGE_USAGE_TRANSIENT_ATTACHMENT_BIT = 0x00000040,
    VK_IMAGE_USAGE_INPUT_ATTACHMENT_BIT = 0x00000080,
    VK_IMAGE_USAGE_FLAG_BITS_MAX_ENUM = 0x7FFFFFFF
} VkImageUsageFlagBits;
typedef VkFlags VkImageUsageFlags;
```

Would you like the header to use the same scheme as Vulkan with a `wgpuFlags` type?

Reply:
ah, hmm. We *do* need some way to disambiguate real "enum" things from bitflags, so it sounds like we need both a "Flags" suffix and a separate `wgpuFlags` typedef...

Reply:
Do we care about potential collisions between bitflags and enums? The only one I can think of is if we add a `WGPUShaderStage` enum at some point.

Reply:
It's just cleaner on the API side when you can instantly see if stuff is supposed to be added together, or only unique value is expected.

Reply:
I thought you said the convention in Rust is to not have any suffix for bitfields? I think I'd be ok keeping the `Bit` suffix as such:
```c
typedef enum WGPUTextureUsageBit {
    WGPUTextureUsageBit_Foo = 0x00000001
    //    VK_IMAGE_USAGE_FLAG_BITS_MAX_ENUM = 0x7FFFFFFF
} WGPUTextureUsageBit;
typedef WGPUFlags WGPUTextureUsageBits;
```

and having the following in C++ (with special sauce bitmaskable enum templates):

```cpp
namespace wgpu {
enum class TextureUsageBit : uint32_t {
    Foo = 0x00000001
};
}
```

Reply:
It's fine for rust because it's a separate type :) so you can't mix up enums with flags.
But for cases (like C) where it's the same type, Flags suffix appears necessary.

Reply:
`WGPUTextureUsageBit_Foo` - can it be `WGPUTextureUsage_Foo` instead?

Reply:
I'm not sure I understand, you'd like to type to have `Bit` or `Bits` but not the enum names?. Can you write down the code you're thinking about for a sample bit flag?

Reply:
The shortest approach would be
```cpp
typedef enum WGPUTextureUsage {
    WGPUTextureUsage_Foo = 0x00000001
} WGPUTextureUsage;
typedef WGPUFlags WGPUTextureUsageFlags;
```

Do you think that could work for us?

Reply:
Ah thanks for the explanation, yeah I think that would work, I'll change it.